Wikipedia/Code/WMFArticleFetcher.m

```diff
@@ -108,26 +134,45 @@ - (nullable NSURLSessionTask *)fetchArticleForURL:(NSURL *)articleURL
+            [taskGroup leave];
         }
         failure:^(NSURLSessionDataTask *operation, NSError *error) {
-            if ([url isEqual:[NSURL wmf_mobileAPIURLForURL:articleURL]] && [error wmf_shouldFallbackToDesktopURLError]) {
-                [self fetchArticleForURL:articleURL useDesktopURL:YES progress:progress failure:failure success:success];
-            } else {
-                [[MWNetworkActivityIndicatorManager sharedManager] pop];
-                failure(error);
-            }
+            articleError = error;
+            [taskGroup leave];
         }];
+
     operation.priority = NSURLSessionTaskPriorityHigh;
     [self trackOperation:operation forArticleURL:articleURL];
+
+    [taskGroup waitInBackgroundAndNotifyOnQueue:dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0)
+                                      withBlock:^{
+                                          [[MWNetworkActivityIndicatorManager sharedManager] pop];
+                                          if (articleResponse && [articleResponse isKindOfClass:[NSDictionary class]]) {
+                                              if (!articleResponse[@"coordinates"] && summaryResponse[@"coordinates"]) {
+                                                  NSMutableDictionary *mutableArticleResponse = [articleResponse mutableCopy];
+                                                  mutableArticleResponse[@"coordinates"] = summaryResponse[@"coordinates"];
+                                                  articleResponse = mutableArticleResponse;
+                                              }
+                                              MWKArticle *article = [self serializedArticleWithURL:articleURL response:articleResponse];
+                                              [self.dataStore asynchronouslyCacheArticle:article];
+                                              dispatch_async(dispatch_get_main_queue(), ^{
+                                                  [self.previewStore addPreviewWithURL:articleURL updatedWithArticle:article];
+                                                  success(article);
+                                              });
+                                          } else {
+                                              if (!articleError) {
+                                                  articleError = [NSError wmf_errorWithType:WMFErrorTypeUnexpectedResponseType userInfo:@{}];
+                                              }
+                                              if ([url isEqual:[NSURL wmf_mobileAPIURLForURL:articleURL]] && [articleError wmf_shouldFallbackToDesktopURLError]) {
+                                                  [self fetchArticleForURL:articleURL useDesktopURL:YES progress:progress failure:failure success:success];
+                                              } else {
+                                                  [[MWNetworkActivityIndicatorManager sharedManager] pop];
```

Comment:
This chunk looks like copy-paste from the previous code, but is this second `pop` still correct since we pop at the top?